code/modules/hydroponics/compost_bin.dm

```diff
@@ -87,20 +110,92 @@
+
+		// Since the character doesn't know what's in the beaker, I'm assuming it is assuming the beaker is full of pure potassium and pours according to that.
+		for(var/E in B.reagents.reagent_list)
+			var/datum/reagent/R = E
+			switch(R.id)
+				if("potassium")
+					potassium += min(R.volume * proportion, potassium_capacity - potassium)
+				if("ash")
+					potash += min(R.volume * proportion, potash_capacity - potash)
+				if("nutriment")
+					biomass += min(R.volume * proportion, biomass_capacity - biomass)
+				if("plantmatter")
+					biomass += min(R.volume * proportion, biomass_capacity - biomass)
+
+			B.reagents.remove_reagent(R.id, R.volume*proportion)
 
-	to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
+		if(proportion == 1)
+			to_chat(user, "<span class='info'>You empty [B] into [src].</span>")
+		else
+			to_chat(user, "<span class='info'>You pour some of [B] into [src].</span>")
+		if(potassium == potassium_capacity)
+			to_chat(user, "<span class='info'>You have saturated the contents of [src] with potassium.</span>")
+		else if(potassium >= potassium_capacity * 0.95)
+			to_chat(user, "<span class='info'>You have very nearly saturated the contents of [src] with potassium.</span>")
+
+		if(potash == potash_capacity)
+			to_chat(user, "<span class='info'>[src] has been filled with potash.</span>")
+		else if(potash >= potash_capacity * 0.95)
+			to_chat(user, "<span class='info'>[src] has been nearly filled with potash.</span>")
+
+		SStgui.update_uis(src)
+		update_icon_state()
+
+		return TRUE
+
+	to_chat(user, "<span class='warning'>You cannot put this in [src]!</span>")
 
 //Compost compostable material if there is any
 /obj/machinery/compost_bin/process()
-	if(compost >= compost_capacity || biomass <= 0)
+	if((compost >= compost_capacity && potassium <= 0) || biomass <= 0)
 		return
 	process_counter++
 	if(process_counter < 5)
 		return
 	process_counter = 0
-	//converts 20% of the biomass to compost each cycle, unless there isn't enough compost space or there is 10 or less biomass
-	var/conversion_amount = clamp(DECAY * biomass, min(MIN_CONVERSION, biomass), compost_capacity - compost)
-	biomass -= conversion_amount
+	//Converts up to 20% of the biomass to compost each cycle, minimum of 10 converted.
+	//In the presence of potassium will create saltpetre crystals instead. Using at most the amount of biomass that would've been used for compost
+	//And making compost from whatever part of that amount it didn't use.
+	var/conversion_amount = max(DECAY * biomass, min(MIN_CONVERSION, biomass))
+	var/potash_saltpetre_conversion = 0
+	var/potassium_saltpetre_conversion = 0
+	var/used_potassium = 0
+	var/used_potash = 0
+
+	if(potash > 0)
+		potash_saltpetre_conversion = min(conversion_amount, potash * BIOMASS_POTASH_RATIO)
+		used_potash = potash_saltpetre_conversion / BIOMASS_POTASH_RATIO
+		saltpetre += used_potash * 2
```

Comment:
Pull the 2 out as a `#define` so it's more clear that it's intentionally increasing the yield.

Reply:
done